pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go

```diff
@@ -72,6 +73,30 @@ func (g *graph) setupShootWatch(ctx context.Context, informer cache.Informer) er
 	return err
 }
 
+func (g *graph) hasExpectedShootBindingEdges(shoot *gardencorev1beta1.Shoot) bool {
+	g.lock.RLock()
+	defer g.lock.RUnlock()
+
+	isEdgeMissingFrom := func(vertexType VertexType, bindingName *string) bool {
+		if bindingName == nil {
+			return false
+		}
+
+		shootVertex, foundShootVertex := g.getVertex(VertexTypeShoot, shoot.Namespace, shoot.Name)
+		bindingVertex, foundBindingVertex := g.getVertex(vertexType, shoot.Namespace, *bindingName)
+		return !foundShootVertex || !foundBindingVertex || !g.graph.HasEdgeFromTo(bindingVertex.id, shootVertex.id)
+	}
+
+	if isEdgeMissingFrom(VertexTypeSecretBinding, shoot.Spec.SecretBindingName) {
+		return false
+	}
+	if isEdgeMissingFrom(VertexTypeCredentialsBinding, shoot.Spec.CredentialsBindingName) {
+		return false
+	}
+
+	return true
```

Comment:
Hm, sorry, now seeing the full result, I think it's readability could be improved even a bit further. WDYT?

```suggestion
	hasEdge := func(vertexType VertexType, bindingName *string) bool {
		if bindingName == nil {
			return true // if bindingName is nil, there is no edge to target vertex, hence we assume the edge exists
		}

		shootVertex, foundShootVertex := g.getVertex(VertexTypeShoot, shoot.Namespace, shoot.Name)
		bindingVertex, foundBindingVertex := g.getVertex(vertexType, shoot.Namespace, *bindingName)
		return foundShootVertex && foundBindingVertex && g.graph.HasEdgeFromTo(bindingVertex.id, shootVertex.id)
	}

	return !hasEdge(VertexTypeSecretBinding, shoot.Spec.SecretBindingName) || !hasEdge(VertexTypeCredentialsBinding, shoot.Spec.CredentialsBindingName)
```

Reply:
accepted after fixing line 90